doc/man/mktap.1
.TH MKTAP "1" "July 2001" "" ""
.SH NAME
mktap \- create twisted.servers
.SH SYNOPSIS
.B mktap
[\fIoptions\fR] \fIapptype\fR [\fIapplication_option\fR]...
.PP
.B mktap
\fIapptype\fR --help
.SH DESCRIPTION
.PP
The \fB\--help\fR prints out a usage message to standard output.
.TP
\fB\--uid\fR, \fB\-u\fR \fI<uid>\fR
Application belongs to this uid, and should run with its permissions.
.TP
\fB\--gid\fR, \fB\-d\fR \fI<gid>\fR
Application belongs to this gid, and should run with its permissions.
.TP
\fB\--append\fR, \fB\-a\fR \fI<file>\fR
Append given servers to given file, instead of creating a new one.
File should be be a tap file.
.TP
\fB\--appname\fR, \fB\-n\fR \fI<name>\fR
Use the specified name as the process name when the application is run with
\fItwistd(1)\fR.  This option also causes some initialization code to be
duplicated when \fItwistd(1)\fR is run.
.TP
\fB\--xml\fR, \fB\-x\fR
Output as a .tax XML file rather than a pickle.
.TP
\fB\--source\fR, \fB\-s\fR
Output as a .tas (AOT Python source) file rather than a pickle.
.TP
\fIapptype\fR
Can be 'web', 'portforward', 'toc', 'coil', 'words', \
'manhole', 'im', 'socks', 'telnet', 'parent', 'sibling', \
'ftp', and 'mail'. Each of those support different options.
.PP
.SH \fBportforward\fR options
.TP
\fB\-h\fR, \fB\--host\fR \fI<host>\fR
Proxy connections to \fI<host>\fR
.TP
\fB\-d\fR, \fB\--dest_port\fR \fI<port>\fR
Proxy connections to \fI<port>\fR on remote host.
.TP
\fB\-p\fR, \fB\--port\fR \fI<port>\fR
Listen locally on \fI<port>\fR
.PP
.SH \fBweb\fR options
.TP
\fB\-u\fR, \fB\--user\fR
Makes a server with ~/public_html and
~/.twistd-web-pb support for users.
.TP
\fB\--personal\fR
Instead of generating a webserver, generate a
ResourcePublisher which listens on ~/.twistd-web-pb
.TP
\fB\--path\fR \fI<path>\fR
<path> is either a specific file or a directory to be
set as the root of the web server. Use this if you
have a directory full of HTML, cgi, php3, epy, or rpy files or
any other files that you want to be served up raw.
.TP
\fB\-p\fR, \fB\--port\fR \fI<port>\fR
<port> is a number representing which port you want to
start the server on.
.TP
\fB\-m\fR, \fB\--mime_type\fR \fI<mimetype>\fR
<mimetype> is the default MIME type to use for
files in a --path web server when none can be determined
for a particular extension. The default is 'text/html'.
.TP
\fB\--allow_ignore_ext\fR
Specify whether or not a request for 'foo' should return 'foo.ext'.
Default is off.
.TP
\fB\--ignore-ext\fR \fI.<extension>\fR
Specify that a request for 'foo' should return 'foo.\fI<extension>\fR'.
.TP
\fB\-t\fR, \fB\--telnet\fR \fI<port>\fR
Run a telnet server on <port>, for additional
configuration later.
.TP
\fB\-i\fR, \fB\--index\fR \fI<name>\fR
Use an index name other than "index.html"
.TP
\fB--https\fR \fI<port>\fR
Port to listen on for Secure HTTP.
.TP
\fB-c\fR, \fB--certificate\fR \fI<filename>\fR
SSL certificate to use for HTTPS. [default: server.pem]
.TP
\fB-k\fR, \fB--privkey\fR \fI<filename>\fR
SSL certificate to use for HTTPS. [default: server.pem]
.TP
\fB--processor\fR \fI<ext>=<class name>\fR
Adds a processor to those file names. (Only usable if after
.B --path)
.TP
\fB--resource-script\fR \fI<script name>\fR
Sets the root as a resource script. This script will be re-evaluated on
every request.
.PP
This creates a web.tap file that can be used by twistd. If you
specify no arguments, it will be a demo webserver that has the Test
class from twisted.web.test in it.
.SH \fBtoc\fR options
.TP
\fB\-p\fR \fI<port>\fR
<port> is a number representing which port you want to
start the server on.
.SH \fBmail\fR options
.TP
\fB\-r\fR, \fB\--relay\fR \fI<ip>,<port>=<queue directory>\fR
Relay mail to all unknown domains through given IP and port,
using queue directory as temporary place to place files.
.TP
\fB\-d\fR, \fB\--domain\fR \fI<domain>\fR=\fI<path>\fR
generate an SMTP/POP3 virtual maildir domain named "domain" which saves to
"path"
.TP
\fB\-u\fR, \fB\--username\fR \fI<name>\fR=\fI<password>\fR
add a user/password to the last specified domains
.TP
\fB\-b\fR, \fB\--bounce_to_postmaster\fR
undelivered mails are sent to the postmaster, instead of being rejected.
.TP
\fB\-p\fR, \fB\--pop\fR \fI<port>\fR
<port> is a number representing which port you want to
start the pop3 server on.
.TP
\fB\-s\fR, \fB\--smtp\fR \fI<port>\fR
<port> is a number representing which port you want to
start the smtp server on.
.PP
This creates a mail.tap file that can be used by twistd(1)
.SH \fBtelnet\fR options
.TP
\fB\-p\fR, \fB\--port\fR \fI<port>\fR
Run the telnet server on <port>
.TP
\fB\-u\fR, \fB\--username\fR \fI<name>\fR
set the username to <name>
.TP
\fB\-w\fR, \fB\--password\fR \fI<password>\fR
set the password to <password>
.SH \fBsocks\fR options
.TP
\fB\-i\fR, \fB\--interface\fR \fI<interface>\fR
Listen on interface <interface>
.TP
\fB\-p\fR, \fB\--port\fR \fI<port>\fR
Run the SOCKSv4 server on <port>
.TP
\fB\-l\fR, \fB\--log\fR \fI<filename>\fR
log connection data to <filename>
.SH \fBftp\fR options
.TP
\fB\-a\fR, \fB\--anonymous\fR
Allow anonymous logins
.TP
\fB\-3\fR, \fB\--thirdparty\fR
Allow third party connections
.TP
\fB\--otp\fR
Use one time passwords (OTP)
.TP
\fB\-p\fR, \fB\--port\fR \fI<port>\fR
Run the FTP server on <port>
.TP
\fB\-r\fR, \fB\--root\fR \fI<path>\fR
Define the local root of the FTP server
.TP
\fB\--anonymoususer\fR \fI<username>\fR
Define the the name of the anonymous user
.SH \fBmanhole\fR options
.TP
\fB\-p\fR, \fB\--port\fR \fI<port>\fR
Run the manhole server on <port>
.TP
\fB\-u\fR, \fB\--user\fR \fI<name>\fR
set the username to <name>
.TP
\fB\-w\fR, \fB\--password\fR \fI<password>\fR
set the password to <password>
.SH \fBwords\fR options
.TP
\fB\-p\fR, \fB\--port\fR \fI<port>\fR
Run the Words server on <port>
.TP
\fB\-i\fR, \fB\--irc\fR \fI<port>\fR
Run IRC server on port <port>
.TP
\fB\-w\fR, \fB\--web\fR \fI<port>\fR
Run web server on port <port>
.SH AUTHOR
Written by Moshe Zadka, based on mktap's help messages
.SH "REPORTING BUGS"
Report bugs to <twisted-python@twistedmatrix.com>.
.SH COPYRIGHT
Copyright \(co 2000 Matthew W. Lefkowitz
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
twistd(1)


NEWS
Ticket numbers in this file can be looked up by visiting
http://twistedmatrix.com/trac/ticket/<number>

Twisted Core 16.0.0pre2 (2016-02-28)
====================================

Features
--------
 - twisted.cred.strports has been ported to Python 3. (#8216)

Deprecations and Removals
-------------------------
 - The __version__ attribute of former subprojects (conch, mail,
   names, news, pair, runner, web, and words) is deprecated in
   preference to the central twisted.__version__. (#8219)

Other
-----
 - #8220


Twisted Conch 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted Mail 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Names 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted News 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Pair 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Runner 16.0.0pre2 (2016-02-28)
======================================

No significant changes have been made for this release.


Twisted Web 16.0.0pre2 (2016-02-28)
===================================

No significant changes have been made for this release.


Twisted Words 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted Core 16.0.0pre1 (2016-02-16)
====================================

Features
--------
 - todo parameter for IReporter.addExpectedSuccess and
   IReporter.addUnexpectedSuccess is no longer required. If not
   provided, a sensible default will be used instead. (#4811)
 - A new string endpoint type, "tls:", allows for properly-verified
   TLS (unlike "ssl:", always matching hostname resolution with
   certificate hostname verification) with faster IPv4/IPv6
   connections.  This comes with an accompanying function,
   twisted.internet.endpoints.wrapClientTLS, which can wrap an
   arbitrary client endpoint with client TLS. (#5642)
 - twisted.python.filepath.makedirs accepts an ignoreExistingDirectory
   flag which ignore the OSError raised by os.makedirs if requested
   directory already exists. (#5704)
 - twisted.protocols.amp has been ported to Python 3. (#6833)
 - twisted.internet.ssl.trustRootFromCertificates returns an object
   suitable for use as trustRoot= to
   twisted.internet.ssl.optionsForClientTLS that trusts multiple
   certificates. (#7671)
 - twisted.python.roots is now ported to Python 3. (#8131)

Bugfixes
--------
 - Expected failures from standard library unittest no longer fail
   with Trial reporters. (#4811)
 - twisted.internet.endpoints.HostnameEndpoint.connect no longer fails
   with an AlreadyCalledError when the Deferred it returns is
   cancelled after all outgoing connection attempts have been made but
   none have yet succeeded or failed. (#8014)
 - twisted.internet.task.LoopingCall.withCount when run with internal
   of 0, now calls the countCallable with 1, regardless of the time
   passed between calls. (#8125)

Improved Documentation
----------------------
 - The Twisted Project has adopted the Contributor Covenant as its
   Code of Conduct. (#8173)

Deprecations and Removals
-------------------------
 - twisted.internet.task.LoopingCall.deferred is now deprecated. Use
   the deferred returned by twisted.internet.task.LoopingCall.start()
   (#8116)
 - twisted.internet.gtkreactor, the GTK+ 1 reactor deprecated since
   Twisted 10.1, has been removed. This does not affect the GTK2,
   GLib, GTK3, or GObject-Introspection reactors. (#8145)
 - twisted.protocols.mice, containing a Logitech MouseMan serial
   driver, has been deprecated. (#8148)

Other
-----
 - #6842, #6978, #7668, #7791, #7881, #7943, #7944, #8050, #8104,
   #8115, #8119, #8122, #8139, #8144, #8154, #8162, #8180, #8187


Twisted Conch 16.0.0pre1 (2016-02-16)
=====================================

Features
--------
 - twisted.conch now uses cryptography instead of PyCrypto for its
   underlying crypto operations. (#7413)
 - twisted.conch.ssh.keys is now ported to Python 3. (#7998)

Bugfixes
--------
 - twisted.conch.ssh.channel.SSHChannel's getPeer and getHost methods
   now return an object which provides IAddress instead of an old-
   style tuple address. (#5999)
 - twisted.conch.endpoint.SSHCommandClientEndpoint, when
   authentication is delegated to an SSH agent, no longer leaves the
   agent connection opened when connection to the server is lost.
   (#8138)

Other
-----
 - #7037, #7715, #8200, #8208


Twisted Mail 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Names 16.0.0pre1 (2016-02-16)
=====================================

No significant changes have been made for this release.


Twisted News 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Pair 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Runner 16.0.0pre1 (2016-02-16)
======================================

No significant changes have been made for this release.


Twisted Web 16.0.0pre1 (2016-02-16)
===================================

Features
--------
 - twisted.web.http_headers._DictHeaders now correctly handles
   updating via keyword arguments in Python 3 (therefore
   twisted.web.http_headers is now fully ported to Python 3). (#6082)
 - twisted.web.wsgi has been ported to Python 3. (#7993)
 - twisted.web.http_headers.Headers now accepts both Unicode and
   bytestring keys and values, encoding to iso-8859-1 and utf8
   respectively. (#8129)
 - twisted.web.vhost ported to Python 3. (#8132)

Bugfixes
--------

twisted/scripts/mktap.py
# Copyright (c) 2001-2004 Twisted Matrix Laboratories.
# See LICENSE for details.

import sys, os

from zope.interface import implements

from twisted.application import service, app
from twisted.persisted import sob
from twisted.python import usage, util, plugin as oldplugin
from twisted import plugin as newplugin

# API COMPATIBILITY
IServiceMaker = service.IServiceMaker


import warnings

warnings.warn(
    "mktap is obsolete as of Twisted 2.5, and officially deprecated "
    "in Twisted 8.0. Use Twisted Application Plugins with the "
    "'twistd' command  directly, as described in "
    "'Writing a Twisted Application Plugin for twistd' chapter of the "
    "Developer Guide.", DeprecationWarning)

try:
    import pwd, grp
except ImportError:
    def getid(uid, gid):
        if uid is not None:
            uid = int(uid)
        if gid is not None:
            gid = int(gid)
        return uid, gid
else:
    def getid(uid, gid):
        if uid is not None:
            try:
                uid = int(uid)
            except ValueError:
                uid = pwd.getpwnam(uid)[2]
        if gid is not None:
            try:
                gid = int(gid)
            except ValueError:
                gid = grp.getgrnam(gid)[2]
        return uid, gid


def loadPlugins(debug = None, progress = None):
    tapLookup = {}

    plugins = oldplugin._getPlugIns("tap", debug, progress)
    for plug in plugins:
        if hasattr(plug, 'tapname'):
            shortTapName = plug.tapname
        else:
            shortTapName = plug.module.split('.')[-1]
        tapLookup[shortTapName] = plug

    plugins = newplugin.getPlugins(IServiceMaker)
    for plug in plugins:
        tapLookup[plug.tapname] = plug

    return tapLookup

def addToApplication(ser, name, append, procname, type, encrypted, uid, gid):
    if append and os.path.exists(append):
        a = service.loadApplication(append, 'pickle', None)
    else:
        a = service.Application(name, uid, gid)
    if procname:
        service.IProcess(a).processName = procname
    ser.setServiceParent(service.IServiceCollection(a))
    sob.IPersistable(a).setStyle(type)
    passphrase = app.getSavePassphrase(encrypted)
    if passphrase:
        append = None
    sob.IPersistable(a).save(filename=append, passphrase=passphrase)

class FirstPassOptions(usage.Options):
    synopsis = """Usage:    mktap [options] <command> [command options] """

    recursing = 0
    params = ()

    optParameters = [
        ['uid', 'u', None, "The uid to run as."],
        ['gid', 'g', None, "The gid to run as."],
        ['append', 'a', None,
         "An existing .tap file to append the plugin to, rather than "
         "creating a new one."],
        ['type', 't', 'pickle',
         "The output format to use; this can be 'pickle', 'xml', "
         "or 'source'."],
        ['appname', 'n', None, "The process name to use for this application."]
    ]

    optFlags = [
        ['encrypted', 'e', "Encrypt file before writing "
                           "(will make the extension of the resultant "
                           "file begin with 'e')"],
        ['debug', 'd',     "Show debug information for plugin loading"],
        ['progress', 'p',  "Show progress information for plugin loading"],
        ['help', 'h',  "Display this message"],
    ]
    #zsh_altArgDescr = {"foo":"use this description for foo instead"}
    #zsh_multiUse = ["foo", "bar"]
    #zsh_mutuallyExclusive = [("foo", "bar"), ("bar", "baz")]
    zsh_actions = {"append":'_files -g "*.tap"',
                   "type":"(pickle xml source)"}
    zsh_actionDescr = {"append":"tap file to append to", "uid":"uid to run as",
                       "gid":"gid to run as", "type":"output format"}

    def init(self, tapLookup):
        sc = []
        for (name, module) in tapLookup.iteritems():
            if IServiceMaker.providedBy(module):
                sc.append((
                    name, None, lambda m=module: m.options(), module.description))
            else:
                sc.append((
                    name, None, lambda obj=module: obj.load().Options(),
                    getattr(module, 'description', '')))

        sc.sort()
        self.subCommands = sc

    def parseArgs(self, *rest):
        self.params += rest

    def _reportDebug(self, info):
        print 'Debug: ', info

    def _reportProgress(self, info):
        s = self.pb(info)
        if s:
            print '\rProgress: ', s,
        if info == 1.0:
            print '\r' + (' ' * 79) + '\r',

    def postOptions(self):
        if self.recursing:
            return
        debug = progress = None
        if self['debug']:
            debug = self._reportDebug
        if self['progress']:
            progress = self._reportProgress
            self.pb = util.makeStatBar(60, 1.0)
        try:
            self.tapLookup = loadPlugins(debug, progress)
        except IOError:
            raise usage.UsageError("Couldn't load the plugins file!")
        self.init(self.tapLookup)
        self.recursing = 1
        self.parseOptions(self.params)
        if not hasattr(self, 'subOptions') or self['help']:
            raise usage.UsageError(str(self))
        if hasattr(self, 'subOptions') and self.subOptions.get('help'):
            raise usage.UsageError(str(self.subOptions))
        if not self.tapLookup.has_key(self.subCommand):
            raise usage.UsageError("Please select one of: "+
                                   ' '.join(self.tapLookup))


def run():
    options = FirstPassOptions()
    try:
        options.parseOptions(sys.argv[1:])
    except usage.UsageError, e:
        print e
        sys.exit(2)
    except KeyboardInterrupt:
        sys.exit(1)

    plg = options.tapLookup[options.subCommand]
    if not IServiceMaker.providedBy(plg):
        plg = plg.load()
    ser = plg.makeService(options.subOptions)
    addToApplication(ser,
                     options.subCommand, options['append'], options['appname'],
                     options['type'], options['encrypted'],
                     *getid(options['uid'], options['gid']))

from twisted.python.reflect import namedAny
from twisted.plugin import IPlugin

class _tapHelper(object):
    """
    Internal utility class to simplify the definition of \"new-style\"
    mktap plugins based on existing, \"classic\" mktap plugins.
    """

    implements(IPlugin, IServiceMaker)

    def __init__(self, name, module, description, tapname):
        self.name = name
        self.module = module
        self.description = description
        self.tapname = tapname

    def options():
        def get(self):
            return namedAny(self.module).Options
        return get,
    options = property(*options())

    def makeService():
        def get(self):
            return namedAny(self.module).makeService
        return get,
    makeService = property(*makeService())
